packages/nouns-contracts/contracts/governance/data/NounsDAOData.sol

```diff
@@ -0,0 +1,272 @@
+            slug,
+            keccak256(encodedProp)
+        );
+    }
+
+    /**
+     * @notice Update a proposal candidate by emitting an event. Nouners can update for free, while non-Nouners must pay `updateCandidateCost` in ETH.
+     * @dev Reverts if the proposer (msg.sender) has not already created a candidate with the same slug.
+     * @param targets the candidate proposal targets.
+     * @param values the candidate proposal values.
+     * @param signatures the candidate proposal signatures.
+     * @param calldatas the candidate proposal calldatas.
+     * @param description the candidate proposal description.
+     * @param slug the candidate proposal slug string, used alognside the proposer to uniquely identify candidates.
+     */
+    function updateProposalCandidate(
+        address[] memory targets,
+        uint256[] memory values,
+        string[] memory signatures,
+        bytes[] memory calldatas,
+        string memory description,
+        string memory slug
+    ) external payable {
+        if (!isNouner(msg.sender) && msg.value < updateCandidateCost) revert MustBeNounerOrPaySufficientFee();
+        if (!propCandidates[msg.sender][keccak256(bytes(slug))]) revert SlugDoesNotExist();
+
+        bytes memory encodedProp = NounsDAOV3Proposals.calcProposalEncodeData(
+            msg.sender,
+            NounsDAOV3Proposals.ProposalTxs(targets, values, signatures, calldatas),
+            description
+        );
+
+        emit ProposalCandidateUpdated(
+            msg.sender,
+            targets,
+            values,
+            signatures,
+            calldatas,
+            description,
+            slug,
+            keccak256(encodedProp)
+        );
+    }
+
+    /**
+     * @notice Cancel a proposal candidate by emitting an event.
+     * @dev Reverts if the proposer (msg.sender) has not already created a candidate with the same slug.
+     * @param slug the candidate proposal slug string, used alognside the proposer to uniquely identify candidates.
+     */
+    function cancelProposalCandidate(string memory slug) external {
+        if (!propCandidates[msg.sender][keccak256(bytes(slug))]) revert SlugDoesNotExist();
+
+        emit ProposalCandidateCanceled(msg.sender, slug);
+    }
+
+    /**
+     * @notice Add a signature supporting a new candidate to be proposed to the DAO using `proposeBySigs`, by emitting an event with the signature data.
+     * @param signer the signer's address.
+     * @param sig the signature bytes.
+     * @param expirationTimestamp the signature's expiration timestamp.
+     * @param encodedPropHash the keccak256 hash of the candidate version signed; the hash is performed on the output of
+     * the `NounsDAOV3Proposals.calcProposalEncodeData` function.
+     */
+    function addSignature(
+        address signer,
+        bytes memory sig,
+        uint256 expirationTimestamp,
+        bytes32 encodedPropHash,
+        string memory reason
+    ) external {
```

Comment:
@eladmallel
1. do we want to add signature validation here?
2. should this also specify the slug?

Reply:
as discussed:

1. leaving out for now, to save on gas 
2. the slug is part of the function and the event it emits 
3. will do to minimize spam

Reply:
ended up doing signature validation because it's the only way of making sure the signer is indeed the account submitting the signature.